### PR TITLE
DEX-946 Order book routes

### PR DESCRIPTION
**Pull request by ngadiyak**

### CODE REVIEW

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/RawHttpChecks.scala [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    r.response.code should be(StatusCodes.MovedPermanently)
+
+  protected def validateMatcherError[ErrorT, EntityT](r: EnrichedResponse[ErrorT, EntityT], c: Int, e: Int, m: String): Unit = {
+    r.response.code should be(c)
+    r.response.headers should contain("Content-Type", "application/json")
+    r.response.body should be leftSideValue
+
+    val b = Json.parse(r.response.body.left.get)
+    (b \ "message").as[String] should be(m)
+    (b \ "error").as[Int] should be(e)
```
Comment (vsuharnikov) [inline dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/RawHttpChecks.scala:25]: 1. Let's get better names for c, e and m
2. You could do the same thing with less efforts:
    ```scala
    protected def validateMatcherError[EntityT](r: EnrichedResponse[MatcherError, EntityT],... = {
      // ...
      val b = r.tryGet.left.get
    ```

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/AsyncEnrichedDexApi.scala [hunk 2]
```diff
@@ -199,6 +199,12 @@ class AsyncEnrichedDexApi(apiKey: String, host: => InetSocketAddress)(implicit e
       .followRedirects(true)
   }
 
+  override def orderBookInfo(amount: String, price: String): R[HttpOrderBookInfo] = mk {
```
Comment (vsuharnikov) [inline dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/AsyncEnrichedDexApi.scala:202]: Let's get better names for arguments:
* amount -> amountAsset
* price -> priceAsset

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/AsyncEnrichedDexApi.scala [hunk 3]
```diff
@@ -199,6 +199,12 @@ class AsyncEnrichedDexApi(apiKey: String, host: => InetSocketAddress)(implicit e
...
+    sttp
+      .get(uri"$apiUri/matcher/orderbook/$amount/$price}/info")
+      .followRedirects(false)
+  }
+
   override def orderBookInfo(assetPair: AssetPair): R[HttpOrderBookInfo] = mk {
```
Comment (vsuharnikov) [inline dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/AsyncEnrichedDexApi.scala:208]: We could use `orderBookInfo(String, String)`:
```scala
override def orderBookInfo(assetPair: AssetPair): R[HttpOrderBookInfo] = 
  orderBookInfo(assetPair.amountAssetStr, assetPair.priceAssetStr)
```

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+          r.headers should contain("Content/type", "application/json")
+
+          val b = Json.parse(r.body.left.toString)
+
+          if (e) {
+            (b \ "message").as[String] should be equals m
+            (b \ "error").as[Int] should be equals mc
+          } else (b \ "status").as[String] should be equals m
+        }
+      }
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala:59]: Positive and negative tests should be separated to make test easier to read

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+      "get tickers for unavailable asset should produce error" in {
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala:61]: Let's replace 'unavailable' with 'unknown'

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        dex1.tryApi.orderBookStatus(wctUsdPair) should failWith(11534345, MatcherError.Params(assetId = Some(WctId.toString)))
+      }
+
+      "status of empty orderbook" in {
+        dex1.api.orderBookInfo(wavesUsdPair).matchingRules.tickSize shouldBe 0.01
+      }
+
+      "error of non-existed order" in {
+        dex1.httpApi.orderBookInfo(AssetPair(usd, Waves)).code shouldBe StatusCodes.MovedPermanently
+      }
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala:71]: `orderBookInfo` != `orderBookStatus`. Suggest:
1. Rename `getMarketStatus` to `getOrderBookStatus` in `MatcherApiRoute`
2. Rename `HttpMarketStatus` to `HttpOrderBookStatus`
3. Move tests with `orderBookInfo` to `GetOrderBookInfoSpec`

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+        val bidAmount = 10000000
+
+        dex1.api.place(mkOrder(bob, bob2WavesPair, SELL, askAmount, ask))
+
+        val resp1 = dex1.api.orderBookStatus(bob2WavesPair)
+        resp1.lastTrade shouldBe None
+        resp1.bestBid shouldBe None
+        resp1.bestAsk should matchTo {
+          Option(LevelAgg(askAmount, ask))
+        }
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala:87]: Or just:
```scala
dex1.api.orderBookStatus(bob2WavesPair) matchTo HttpOrderBookStatus(
  // ...
)
```

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        dex1.api.place(mkOrder(alice, bob2WavesPair, BUY, bidAmount, bid))
+
+        val resp2 = dex1.api.orderBookStatus(bob2WavesPair)
+        resp2.lastTrade should matchTo {
+          Option(LastTrade(ask, askAmount, BUY))
+        }
+        resp2.bestBid should matchTo {
+          Option(LevelAgg(bidAmount - askAmount, bid))
+        }
+        resp2.bestAsk shouldBe None
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala:98]: Same here

**Review by vsuharnikov — CHANGES_REQUESTED** [review]

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/AsyncEnrichedDexApi.scala (hunk 2 above)
Comment (ngadiyak) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/AsyncEnrichedDexApi.scala:202]: ok

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/AsyncEnrichedDexApi.scala (hunk 3 above)
Comment (ngadiyak) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/dex/AsyncEnrichedDexApi.scala:208]: ok

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+      withClue("- trade (best sell order partially filled)") {
+        placeAndAwaitAtNode(mkOrder(alice, wavesUsdPair, BUY, 1.waves, 2.usd))
+        val status = validate200Json(dex1.rawApi.getOrderBookStatus(wavesUsdPair))
+        status.ask.get should be(1.9.usd)
+        status.askAmount.get should be(1.waves)
+        status.bid.get should be(1.1.usd)
+        status.bidAmount.get should be(10.waves)
+      }
+//        //TODO: will be fixed in the morning, want to sleep ))
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala:76]: 😄

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+//      withClue("- trade (best sell order fully filled)") {
+//        placeAndAwaitAtNode(mkOrder(alice, wavesUsdPair, BUY, 1.waves, 2.usd))
+//        val status = validate200Json(dex1.rawApi.getOrderBookStatus(wavesUsdPair))
+//        status.ask.get should be(1.9.usd)
+//        status.askAmount.get should be(1.waves)
+//        status.bid.get should be(1.1.usd)
+//        status.bidAmount.get should be(1.waves)
+//        status.lastPrice.get should be(LastTrade(1.9.usd, 1.waves, BUY))
+//        status.bestBid.get should be(LevelAgg(1.waves, 1.1.usd))
+//        status.bestAsk.get should be(LevelAgg(1.waves, 1.9.usd))
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala:91]: Use `matchTo HttpOrderBookStatus(`

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/RawHttpChecks.scala [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    r.response.code should be(StatusCodes.MovedPermanently)
+
+  protected def validateMatcherError[ErrorT, EntityT](r: EnrichedResponse[ErrorT, EntityT], code: Int, error: Int, message: String): Unit = {
+    r.response.code should be(code)
+    r.response.headers should contain("Content-Type", "application/json")
+    r.response.body should be leftSideValue
+
+    val b = Json.parse(r.response.body.left.get)
+    (b \ "message").as[String] should be(message)
+    (b \ "error").as[Int] should be(error)
```
Comment (vsuharnikov) [inline dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/RawHttpChecks.scala:25]: What about:
```scala
protected def validateMatcherError[EntityT](r: EnrichedResponse[MatcherError, EntityT],... = {
  // ...
  val b = r.tryGet.left.get
  b should matchTo(MatcherError(// ...
```
?

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookInfoSpec.scala [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+    "should return exception when price is not a correct base58 string" in {
+      validate404Exception(dex1.rawApi.getOrderBookInfo("WAVES", "null"))
+    }
+
+    "should return correct data" in {
+      validate200Json(dex1.rawApi.getOrderBookInfo(wavesUsdPair)).matchingRules should be(HttpMatchingRules(0.01))
+    }
+
+    forAll(negativeAssets) { (a: String, p: String, c: Int, e: Int, m: String) =>
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookInfoSpec.scala:44]: Please inline `negativeAssets` to `forAll` to see what `a`, `p`, `c`, `e` and `m` means

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+      cancelAndAwait(alice, ask)
+
+      validate200Json(dex1.rawApi.getOrderBook(wavesUsdPair)).asks should have size 0
+    }
+
+    "should return exception when price is not a correct base58 string" in {
+      validate404Exception(dex1.rawApi.getOrderBook("WAVES", "null"))
+    }
+
+    forAll(negativeDepth) { (v: String, t: String) =>
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:70]: Same here with inlining

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+      s"for depth = $v ($t) should return exception" in { // TODO fix, endpoint should return something like 400/Bad Request
+        validate404Exception(dex1.rawApi.getOrderBook(wavesUsdPair, v))
+      }
+    }
+
+    forAll(negativeAssets) { (a: String, p: String, c: Int, e: Int, m: String) =>
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:76]: And here too

File: dex-it/src/test/scala/com/wavesplatform/it/api/MatcherState.scala [hunk 15]
```diff
@@ -5,13 +5,13 @@ import com.wavesplatform.dex.domain.account.KeyPair
...
-  reservedBalances: Map[KeyPair, HttpBalance],
-  orderHistory: Map[KeyPair, Map[AssetPair, Seq[HttpOrderBookHistoryItem]]]
+                         offset: HttpOffset,
+                         snapshots: HttpSnapshotOffsets,
+                         orderBooks: Map[AssetPair, (HttpV0OrderBook, HttpOrderBookStatus)],
+                         orderStatuses: Map[String, HttpOrderStatus],
+                         orderTransactionIds: Map[String, Set[String]],
+                         reservedBalances: Map[KeyPair, HttpBalance],
+                         orderHistory: Map[KeyPair, Map[AssetPair, Seq[HttpOrderBookHistoryItem]]]
 ) {
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/api/MatcherState.scala:15]: Probably a formatting is broken. Try to reformat the code

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+      r.asks should have size 0
+      r.bids should have size 0
+    }
+
+    "should return correct number of asks and bids" in {
+      val ask = mkOrder(alice, wavesUsdPair, SELL, 10.waves, 5.usd)
+
+      placeAndAwaitAtDex(mkOrder(alice, wavesUsdPair, BUY, 1.waves, 1.usd))
+      placeAndAwaitAtDex(ask)
+      placeAndAwaitAtDex(mkOrder(alice, wavesUsdPair, BUY, 3.waves, 4.usd))
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:53]: A small hint to make slightly less code:
```scala
List(
  mkOrder(alice, wavesUsdPair, BUY, 1.waves, 1.usd), 
  ask,
  mkOrder(alice, wavesUsdPair, BUY, 3.waves, 4.usd)
).foreach(placeAndAwaitAtDex(_))
```

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+
+    "should return exception when amount is not a correct base58 string" in { // TODO: ? Create task for change it to matcherError?
+      validate404Exception(dex1.rawApi.getOrderBookStatus("null", "WAVES"))
+    }
+
+    "should return exception when price is not a correct base58 string" in {
+      validate404Exception(dex1.rawApi.getOrderBookStatus("WAVES", "null"))
+    }
+
+    forAll(negativeAssets) { (a: String, p: String, c: Int, e: Int, m: String) =>
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala:103]: Same here with inlining

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+  override protected def dexInitialSuiteConfig: Config = ConfigFactory.parseString(
+    s"""waves.dex {
+       |  price-assets = [ "$BtcId", "$UsdId", "WAVES" ]
+       |}""".stripMargin
+  )
+
+  override protected def beforeAll(): Unit = {
+    wavesNode1.start()
+    broadcastAndAwait(IssueBtcTx)
+    broadcastAndAwait(IssueUsdTx)
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:34]: If you write:
```scala
broadcastAndAwait(IssueBtcTx, IssueUsdTx)
```
the test will take multiple seconds less :)

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+  private val negativeAssets = Table(
+    ("Amount", "Price", "Http status", "Error code", "Message"),
+    ("incorrect", "WAVES", 404, 11534345, "The asset incorrect not found"),
+    ("WAVES", "incorrect", 404, 9440771, "The WAVES-incorrect asset pair should be reversed")
+  )
+
+  override protected def beforeAll(): Unit = {
+    wavesNode1.start()
+    broadcastAndAwait(IssueBtcTx)
+    broadcastAndAwait(IssueUsdTx)
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala:30]: Same

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBooksSpec.scala [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+  override protected def dexInitialSuiteConfig: Config = ConfigFactory.parseString(
+    s"""waves.dex {
+       |  price-assets = [ "$BtcId", "$UsdId", "WAVES" ]
+       |}""".stripMargin
+  )
+
+  override protected def beforeAll(): Unit = {
+    wavesNode1.start()
+    broadcastAndAwait(IssueBtcTx)
+    broadcastAndAwait(IssueUsdTx)
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBooksSpec.scala:23]: Same

**Review by vsuharnikov — CHANGES_REQUESTED** [review]

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala (hunk 16 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:53]: :+1:

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBooksSpec.scala (hunk 20 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBooksSpec.scala:23]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala (hunk 19 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala:30]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala (hunk 18 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:34]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala (hunk 16 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:53]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookInfoSpec.scala (hunk 12 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookInfoSpec.scala:44]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala (hunk 14 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:76]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala (hunk 17 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala:103]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/api/MatcherState.scala (hunk 15 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/api/MatcherState.scala:15]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala (hunk 13 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookSpec.scala:70]: fixed

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala (hunk 10 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala:91]: ok

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala (hunk 9 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookStatusSpec.scala:76]: :)

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala (hunk 6 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala]: OK!

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala (hunk 4 above)
Comment (ngadiyak) [reply dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetMarketStatusSpec.scala]: fixed

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/RawHttpChecks.scala (hunk 11 above)
Comment (ngadiyak) [reply dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/RawHttpChecks.scala:25]: make attempt to fix, see in code

File: dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/RawHttpChecks.scala [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+
+  protected def validateMatcherError[ErrorT, EntityT](r: EnrichedResponse[ErrorT, EntityT], code: Int, error: Int, message: String): Unit = {
+    r.response.code should be(code)
+    r.response.headers should contain("Content-Type", "application/json")
+    r.response.body should be leftSideValue
+
+    r.tryGet match {
+      case Left(MatcherError(e, m, _, _)) => e should be(error); m should be(message);
+      case _ => assert(false, s"Unexpected response $r")
+    }
```
Comment (vsuharnikov) [inline dex-it-common/src/main/scala/com/wavesplatform/dex/it/api/RawHttpChecks.scala:26]: 👍

File: dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookInfoSpec.scala [hunk 22]
```diff
@@ -1,23 +1,40 @@
...
 
 class GetOrderBookInfoSpec extends MatcherSuiteBase with TableDrivenPropertyChecks with RawHttpChecks {
 
+  val minAmount = 0.1
+  val maxAmount = 100000000
+  val stepAmount = 0.1
+  val minPrice = 0.0001
+  val maxPrice = 1000
+  val stepPrice = 0.001
+
```
Comment (vsuharnikov) [inline dex-it/src/test/scala/com/wavesplatform/it/matcher/api/http/GetOrderBookInfoSpec.scala:18]: You can create `expectedOrderRestrictions : HttpOrderRestrictions` here and import it in `dexInitialSuiteConfig`:
```scala
override protected def dexInitialSuiteConfig: Config = {
  import expectedOrderRestrictions._
  // ...
}
```

**Review by vsuharnikov — APPROVED** [review]

**Comment by vsuharnikov** [comment]
approved

**Pull request merged**